WordAds: Update EU Cookie Law Widget w/ ads compliance features

rclations opened the pull request:
#### Changes proposed in this Pull Request:

* Add new setting for:
  * Banner consent expiration. 
_This updates the current default (30 days) to be a user-specified setting. New default is 180 days (6mo)._
* Add new filter `jetpack_disable_eu_cookie_law_widget`, which can be used to disable the banner.
  * Usage example: a site using the [MaxMind GeoIP database](https://www.maxmind.com/en/geoip-demo) on their server could write a filter that extrapolates each visitor's country from their IP and uses that to show banners for users from certain countries.
* Adds requirement (+ notice) that explicit button-click opt-in for ads module users.
* Adds a new `personalized-ads-consent` cookie for ads-enabled sites. 
The cookie banner will fire if either cookie is missing, and a consent action will update both cookies with the new date value. This method will cover use cases where users gain consent from users, but later turn on ads (which they may not have consent to personalize).

* Additional UI changes
  * Cherry-picked the changes from #9559 & #9463 which also update some language, so we can get a holistic review on this PR. Most notable is that the widget has been renamed for clarity.

  * Reordered the widget form sections in a more intuitive way. (Banner display > consent options > coloring)

  * Updated the default link text to be a part of the widget form. Previously, To find out more, including how to control cookies, see here: would be pre-pended to the widget link, but there was no way for the user to control this text. This seemed confusing to me when we had a textarea for that purpose.

  * Updated the widget output to allow users to use line breaks in the banner text field (just as we are in the default text).

  * Updated the widget form section labels to address some of the clarity issues you mentioned above.


<img width="315" alt="screen shot 2018-05-17 at 11 31 08 am" src="https://user-images.githubusercontent.com/2522431/40187575-d6bc54d2-59c5-11e8-9e55-d43674c91bf7.png">

#### Testing instructions:

* Test the new settings, make sure they behave as expected.
* Enable the Ads module, then click the consent banner. Make sure the `personalized-ads-consent` cookie fires.

#### Additional Notes:

* This will require updating https://jetpack.com/support/cookies/ to include ad cookie information (as well as the `personalized-ads-consent` cookie introduced here.

bubel commented:
@rclations can you send me the ad cookie details/information? I can get the cookie documentation updated accordingly.

rclations commented:
Switched detection method to the Geolite2 library (also [used by Woo](https://woocommerce.wordpress.com/2018/04/18/woocommerce-3-4-beta-is-here/))

Updated testing instructions.

jeherve commented on _inc/lib/class.jetpack-geolite-integration.php:
```diff
@@ -0,0 +1,70 @@
+<?php
+/**
+ * Wrapper for MaxMind GeoLite2 Reader
```
Discussion to be continued in pMyNb-4ce-p2

imsimond commented on modules/widgets/eu-cookie-law.php:
```diff
@@ -22,11 +22,56 @@ class Jetpack_EU_Cookie_Law_Widget extends WP_Widget {
+		 * @var array
+		 */
+		public static $gdpr_zone = array(
+			// European Member countries
```
More accurately here, European Union member countries

rclations commented:
Removed all geolocation features and leaving 'in progress' for now.

bubel commented on modules/widgets/eu-cookie-law/form.php:
```diff
@@ -13,6 +13,11 @@
 			</label>
+			<?php if ( Jetpack::is_module_active( 'wordads' ) ) : ?>
+				<p>
+					<strong><em><?php esc_html_e( 'This option is strongly recommended for Ads users.', 'jetpack' ); ?></em></strong>
```
Should we be a bit more specific and say "WordAds" (vs. just "Ads") here and maybe link to its support document to better explain the reference?

rclations replied:
Updated to "Jetpack Ads" in the most recent commit to make the reference more clear.

rclations commented:
Updated the PR description + screenshot above with the latest commit.

allendav commented:
Will the privacy policy URL auto populate with the URL corresponding to the core setting for the privacy policy page ID (wp_page_for_privacy_policy) if present?

allendav commented:
As mentioned earlier today, if they elect to link to a custom privacy policy, and they have Word Ads enabled, that we should prompt them to be sure that their privacy policy mentions Word Ads.

rclations commented:
@allendav:
>Will the privacy policy URL auto populate with the URL corresponding to the core setting for the privacy policy page ID (wp_page_for_privacy_policy) if present?

@jaswrks, wanted to check if you'd looked into this on your work in #9336 - thinking this idea might fit an an enhancement to that?

jaswrks commented:
@rclations writes...

> wanted to check if you'd looked into this on your work in #9336 
> thinking this idea might fit an an enhancement to that?

I didn't even think of that. Yes, that'd make an excellent enhancement. It could serve as a new default value for the privacy policy URL in the Jetpack EU cookie widget.

rclations commented:
1. Added a warning notice per @allendav's suggestion above:
>As mentioned earlier today, if they elect to link to a custom privacy policy, and they have Word Ads enabled, that we should prompt them to be sure that their privacy policy mentions Word Ads.

2. Updated the style of the warning notices to match the work done in #9336

I also added some description text below the `Hide the banner` section explaining that this determines how consent is captured. (When writing the ads notice here, it seemed confusing without explaining that piece)

Updated screenshot below:

<img width="358" alt="screen shot 2018-05-11 at 11 44 16 am" src="https://user-images.githubusercontent.com/2522431/39933346-b89e6b70-5510-11e8-845e-e126865d79be.png">

rickybanister commented:
This generally looks good, but I've noticed a few places around Jetpack and our products lately that we will introduce a difficult to understand concept somewhat flippantly and only deeper into the UI do we provide an explanation. 

In the above UI we sort of make the assumption that the customer will understand that part of the banner includes the ability to link to a page that contains a privacy policy, but we don't ever really tell them that. There isn't an over-arching description that explains what all this is about. We even shorten European Union to EU, which may not be obvious to non-eu folks. 

So I suppose I'm recommending adding a little tiny bit of explanatory information and possibly adjusting the order of some things. The title 'policy url' would benefit from having some descriptive text below it that explains what we're talking about. 

I realize that feedback is not relevant to the scope of this PR, I think using the notice style helps draw attention to that information.

rclations commented:
Thanks for the review @rickybanister - this makes a lot of sense.

I've made a few updates:

- Cherry-picked the changes from #9559 & #9463 which also update some language, so we can get a holistic review on this PR. Most notable is that the widget has been renamed for clarity.
- Reordered the widget form sections in a more intuitive way. (Banner display > consent options > coloring)
- Updated the default link text to be a part of the widget form. Previously, `To find out more, including how to control cookies, see here:` would be pre-pended to the widget link, but there was no way for the user to control this text. This seemed confusing to me when we had a textarea for that purpose.
- Updated the widget output to allow users to use line breaks in the banner text field (just as we are in the default text).
- Updated the widget form section labels to address some of the clarity issues you mentioned above.

Here is a screenshot of the new widget form:

<img width="315" alt="screen shot 2018-05-17 at 11 31 08 am" src="https://user-images.githubusercontent.com/2522431/40187575-d6bc54d2-59c5-11e8-9e55-d43674c91bf7.png">

rickybanister commented:
I think that flows a bit better. I wonder though if we might modify the description/warning at the very bottom to be a bit more of an introduction to what all this is for, and then move it to the very top? If someone stumbles on this by accident I'd love for the first thing they read to give them a run down of 'yes this is for you' or 'don't worry this doesn't apply to you' with the link to learn more.

oskosk commented:
Ported to `branch-6.1` in b25f7fb472eea4c47edcf7a2317bcf8e9114c191
